### PR TITLE
[TEST] Add test for LeakTracker

### PR DESCRIPTION
**Pull request by nicktindall**
Related to #108123

Uses `assertLeakDetected` created for testing #109325

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/transport/LeakTrackerTests.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+            new ReleasableTrackedObjectLifecycle(),
+            new ReferenceCountedTrackedObjectLifecycle()
+        ).map(i -> new Object[] { i }).toList();
+    }
+
+    public void testWillLogErrorWhenTrackedObjectIsNotClosed() throws Exception {
+        // Let it go out of scope without closing
+        trackedObjectLifecycle.createAndTrack(reachabilityChecker);
+        reachabilityChecker.ensureUnreachable();
+        assertBusy(() -> assertLeakDetected("LEAK: resource was not cleaned up before it was garbage-collected\\.(.*|\\s)*"));
```
Comment (nicktindall) [inline server/src/test/java/org/elasticsearch/transport/LeakTrackerTests.java:62]: This regex will match any `LeakTracker`-originated leak (it won't match those originating from netty). It'll probably end up being repeated in all such tests. I wonder if it's worth implementing an overload that takes no regex and matches any leaks, if you're testing some kind of leak detection you probably just care that one was detected?

Comment (nicktindall) [reply server/src/test/java/org/elasticsearch/transport/LeakTrackerTests.java:62]: The only way to be more specific, currently, is by matching the stack-trace. We'd need to allow the callers of `LeakTracker#track` to provide something to be included in the message if we wanted to enable more specific matching on the message alone.

Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/transport/LeakTrackerTests.java:62]: I think it's fine to just assert that a leak was detected without needing to be precise about exactly which one it was. We can always revisit later if we need something more discriminating, but I expect that any tests that deliberately generate leaks won't really be the best place to detect other inadvertent leaks too.

File: test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java [hunk 2]
```diff
@@ -790,17 +789,12 @@ protected static void checkStaticState() throws Exception {
...
+     * so the test won't fail for leaks detected up until this point.
      */
-    protected static void assertLeakDetected(String expectedPattern) {
+    protected static void assertLeakDetected() {
         synchronized (loggedLeaks) {
-            assertTrue(
-                "No leak detected matching the pattern: " + expectedPattern,
-                loggedLeaks.removeIf(leakText -> Pattern.matches(expectedPattern, leakText))
-            );
+            assertTrue("No leaks have been detected", loggedLeaks.removeIf(leakText -> true));
```
Comment (DaveCTurner) [inline test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java:797]: nit: seems a little too sneaky, I think I'd prefer a more explicit `assertFalse(empty())` and then `clear()` similar to the bottom of `assertStaticState()` above.

**Review by DaveCTurner — APPROVED** [review]
LGTM (one nit)

**Pull request merged**